### DISCUSSION TITLE
Window Swallowing

### DISCUSSION THREAD

**Discussion by arthsmn** [niri-wm/niri#762] [Ideas]
I would like to request a way to set a regex to swallow windows, like mpv in a terminal. Hyprland has it and it works like a charm, super useful.

**Comment by YaLTeR** [comment]
What does swallowing do?

**Reply by arthsmn** [reply]
Basically whenever you spawn a window from another one, the window is hidden and the spawned one gets in its place. Eg. spawning mpv through a terminal will hide the terminal and place mpv in the same spot as the terminal, when it closes, the terminal reappears at the same spot.

Example video: https://youtu.be/92uo5OBOKfY?feature=shared

**Reply by YaLTeR** [reply]
I see. Well, this is not something that the niri layout does currently, so it would likely be quite a bit of work. I'm not sure it's useful enough to warrant it?

**Reply by arthsmn** [reply]
> I see. Well, this is not something that the niri layout does currently, so it would likely be quite a bit of work. I'm not sure it's useful enough to warrant it?

I use it on Hyprland to run apps from terminals and other programs like image viewers from apps, but I know it's not something high priority. Thanks for the quick response and hope someday this gets implemented :)

**Comment by nakibrayane** [comment]
This is not useful in niri, the main reason for window swallowing is to not have the terminal window waste screen space, but in niri the columns are infinitely big, so you are not wasting any space by leaving the terminal window open, you could still have a full width mpv window and a full width terminal window in the same workspace

**Reply by arthsmn** [reply]
Yeah but then you would have 2 windows opened, when you could only have mpv (as the terminal isn't useful), wasting a movement.

**Reply by nakibrayane** [reply]
But then when you want to access the terminal for some reason, could you switch back, or you would have to close the opened window (mpv) first?

And would this feature work only on windows you select using a window rule, for example, terminal and mpv, or would it work on all windows? if so how would programs that spawn other windows behave, for example, gnu octave spawns new windows but you still need to access the old window, would the newly opened window swallow the old one?

**Reply by arthsmn** [reply]
> But then when you want to access the terminal for some reason, could you switch back, or you would have to close the opened window (mpv) first?
> 
> And would this feature work only on windows you select using a window rule, for example, terminal and mpv, or would it work on all windows? if so how would programs that spawn other windows behave, for example, gnu octave spawns new windows but you still need to access the old window, would the newly opened window swallow the old one?

Generally you would hide the terminal and place mpv in its place, so you would need to close mpv to interact with the terminal (which is not a big deal since the terminal isn't doing much). Every wm I see use regex to match the window title, this way you can do to only certain windows or everything (Hyprland does this for example).

**Reply by panchoh** [reply]
(Hyprland user here).
I also have bindings in place to `unswallow/reswallow` when needed.
Having that flexibility in Niri would be _really_ nice.

**Comment by dogeystamp** [comment]
I hacked together a script that does window swallowing, kind of like [swayhide](https://github.com/rehanzo/swayhide) does:

```bash
#!/bin/sh
# window "swallow" for Niri
# (https://github.com/YaLTeR/niri/discussions/762)
#
# this script shrinks the terminal into a column with the new window,
# incredibly hacky technique
#
# usage: niri-swal.sh mpv video.mp4
#
# useful to bind in your shell.

set -e

TERM_ID=$(niri msg focused-window | head -n 1 | grep -o '[0-9]*')

$@ &
# may need to adjust this
sleep 0.5

niri msg action focus-window --id "$TERM_ID"
niri msg action consume-window-into-column
niri msg action move-window-down
niri msg action set-window-height --id "$TERM_ID" 0%
niri msg action focus-window-up

wait
niri msg action set-window-height --id "$TERM_ID" 100%
```

It's probably highly unreliable, but it seems to work.

**Reply by YaLTeR** [reply]
You can use `focus-window --id`, but for the other two commands I haven't made them by-id yet. Though, in unreleased niri, `consume-or-expel-window-left` takes an `--id`.

**Reply by dogeystamp** [reply]
> You can use `focus-window --id`, but for the other two commands I haven't made them by-id yet. Though, in unreleased niri, `consume-or-expel-window-left` takes an `--id`.

Would it be possible for the `consume` command to take an id for the window
to consume, and not the one consuming?

In the proper, automatic swallowing implementation that [dwm](http://dwm.suckless.org/patches/swallow/) and
[hyprland](https://github.com/hyprwm/Hyprland/commit/a97621b1cb203b0e3a7dd07d88c5422b798cbfd8) use, new windows are swallowed if their process is the descendant
of the terminal's (based on PID, parent PID).

In niri, we could simulate swallowing similar to how dwm's [dynamicswallow](http://dwm.suckless.org/patches/dynamicswallow/)
works. While starting the app to be swallowed, we have a script watch the
event stream and match the new window that appears based on its class, and then
consume that by its ID.

**Reply by YaLTeR** [reply]
consume-or-expel-window-left accepts the window that will be consumed (or expelled). The consume-window-into-column does not accept an id atm, and I'm not sure it should, because it's a column-level action, not a window-level action.

**Reply by dogeystamp** [reply]
I wrote an improved script that uses the new `consume-or-expel-window-left`:
```bash
#!/bin/bash
# window swallow emulation for Niri
# (https://github.com/YaLTeR/niri/discussions/762)
# requires at least commit d5cbc35 / version 0.1.10
#
# this script shrinks the terminal into a column with the new window.
# bind it in your shell as an alias like "hide".
#
# example:
#
# 	niri-swal.sh mpv mpv video.mp4
# 	niri-swal.sh org.pwmt.zathura zathura document.pdf
#
# an app_id argument is required to ensure the right window is being swallowed.
# to get this argument, use `niri msg windows`.

set -e

TERM_ID="$(niri msg --json focused-window | jq -r .id)"
APP_ID="$1"
shift

{
	# if within 7 events, we don't see the window open, give up to avoid swallowing unexpectedly later
	APP_WIN_ID=$(
		# head -n 1 will wait for a second event before terminating
		# https://superuser.com/a/275962
		head -n 1 <(
			niri msg --json event-stream | jq --unbuffered --null-input --raw-output "range(7) as \$i | input | .WindowOpenedOrChanged | select(.) | select (.window.app_id == \"$APP_ID\") | .window.id"
		)
	)

	if [ -z "$APP_WIN_ID" ]; then
		exit 1
	fi

	niri msg action focus-window --id "$TERM_ID"
	niri msg action consume-or-expel-window-left --id "$APP_WIN_ID"
	niri msg action move-window-down
	niri msg action set-window-height --id "$TERM_ID" 0%
	niri msg action focus-window-up
} &

"$@"

niri msg action set-window-height --id "$TERM_ID" 100%

```

![swallow](https://github.com/user-attachments/assets/6ffb02eb-754e-4174-973e-8e9ba30825d8)

There are a few bugs in the script though. Why is the niri event "WindowOpenedOrChanged"? I'm not sure how to distinguish between open and change.
I only want to subscribe to the window open, but sometimes a window changes before that event and it screws up the script.

Another issue with the script is that it can't swallow a window across workspaces. For example, I start `mpv` in a terminal on workspace 1, and before the window opens, I switch to workspace 2.
Then `mpv` appears on workspace 2, and it can't get swallowed because it's not on the same workspace as the terminal.
I tried to move the window programmatically across workspaces, but `move-window-to-workspace` doesn't take the "Workspace ID" that the script can read, only the user-facing workspace number.

Thanks for your input, and also thanks for making such a nice compositor.

**Reply by YaLTeR** [reply]
This looks interesting!

> Why is the niri event "WindowOpenedOrChanged"? I'm not sure how to distinguish between open and change.

The event stream is stateful. You're generally supposed to keep track of the current state. So if the window wasn't previously open, then it was opened, otherwise an existing window changed.

> but move-window-to-workspace doesn't take the "Workspace ID" that the script can read, only the user-facing workspace number

Yeah, #647. You can actually reference it via JSON but not via the CLI right now.

**Comment by YaLTeR** [comment]
> In the proper, automatic swallowing implementation that [dwm](http://dwm.suckless.org/patches/swallow/) and
[hyprland](https://github.com/hyprwm/Hyprland/commit/a97621b1cb203b0e3a7dd07d88c5422b798cbfd8) use, new windows are swallowed if their process is the descendant
of the terminal's (based on PID, parent PID).

I added PID to the Window IPC if it helps in d193928f315912b67cc997ab4e6f96675695374e. However, it seems that toolbox containers (and I imagine distrobox and others too) don't preserve the parent terminal PID:

```
└─ pstree -s 53997
systemd───systemd───conmon───fish───weston-flower
```

So a parent-PID-based solution will fail if you're primarily working in a toolbox (e.g. on an immutable distro).

**Comment by markstos** [comment]
Since 25.02, the "toggle-column-tabbed-display" action was added. This allows swallowing to emulated through this sequence:

1. Set the current column to tabbed.
2. Spawn the new window
3. Wait long enough for the window to appear.
4. Consume window back to the original column.

The effect is that the spawned window ends up on visually top of the old one. When you close it, the old one appears.

Niri could implement this internally by wrapping the steps enough into a slicker "swallow" command that's the same end result without the visual stutter.

Try this proof-of-concept yourself:

```
# Run from an untabbed terminal in a column by itsself.
niri msg action toggle-column-tabbed-display;  \ 
niri msg action spawn -- firefox; \ 
sleep 1;  \ 
niri msg action consume-or-expel-window-left
```

**Comment by dogeystamp** [comment]
Here's an update on the [swallow script](https://github.com/dogeystamp/dots/blob/8299e0b2c0db4d287640e588d075985385fba81e/src/dot_local/bin/executable_niri-swal.sh) I made a year ago. I have updated it to use tabs to emulate swallowing.
I have been using this script regularly, and it works 95% of the time. If you want to have window swallowing, feel free to use it.

https://github.com/user-attachments/assets/86dd63aa-4ab2-413f-9876-03997e16c216


There are two main issues for implementing swallowing using a script.

First, we don't know which window has to be swallowed, and which window is the terminal that will consume it.

Generally, it's not as simple as just "spawn the window and `consume-or-expel-window-left`," because if the user focuses on a different window while the swallowing is happening, the wrong window ends up being consumed. For instance, `mpv` might take 10 seconds to open a window when playing a video over the network, and during this time the user might switch workspace, and then have a random window be consumed left.

We could use PID to identify the window that does the swallowing, but this is not reliable, e.g. because Alacritty terminals can all share the same PID. In my script, I assume that the window focused when the command is started is the window that does the swallowing, and use app id to match the window that is swallowed. This is generally reliable. 

Secondly, it's hard to place an arbitrary window into an arbitrary column programmaticaly in Niri.
We only have "consume left" or "consume into focused column," which doesn't work when the column we want to consume into is on a different workspace. Ideally, we would have a niri action that does "consume window ID 1234 into the column containing window ID 5678."

A lot of code in my swallow script is dedicated to shuffling windows around to consume the correct windows, like this:
```shell
	# move the window to the correct monitor
	niri msg action move-window-to-monitor --id "$APP_WIN_ID" "$TERM_OUTPUT"
	err_printf "\nmoved win %s (APP_WIN_ID) to monitor %s (TERM_OUTPUT)\n" "$APP_WIN_ID" "$TERM_OUTPUT"

	# move the new window onto a random workspace to make it forget its location
	niri msg action move-window-to-workspace --window-id "$APP_WIN_ID" --focus false 255
	err_printf "\nmoved win %s (APP_WIN_ID) to workspace 255\n" "$APP_WIN_ID"

	# move the new window onto the terminal's workspace
	sock_printf '{"Action":{"MoveWindowToWorkspace":{"window_id":%s,"reference":{"Id":%s},"focus":true}}}' "$APP_WIN_ID" "$TERM_WORKSPACE_ID"
	err_printf "\nmoved win %s (APP_WIN_ID) to workspace %s (TERM_WORKSPACE_ID)\n" "$APP_WIN_ID" "$TERM_WORKSPACE_ID"

	# consume the app into the terminal column
	niri msg action consume-or-expel-window-left --id "$APP_WIN_ID"
	niri msg action set-column-display tabbed
	niri msg action focus-window --id "$APP_WIN_ID"

	# focus back onto the workspace the app spawned on
	CURRENT_FOCUSED_WORKSPACE=$(niri msg --json focused-window | jq -r .workspace_id)
	if [ "$CURRENT_FOCUSED_WORKSPACE" != "$WORKSPACE_ID" ]; then
		sock_printf '{"Action":{"FocusWorkspace":{"reference":{"Id":%s}}}}' "$WORKSPACE_ID"
	fi

	# focus back onto the monitor that was focused
	niri msg action focus-monitor "$FOCUSED_OUTPUT"
```
This code is ugly, and it's also brittle. It's not really a high priority since I already implemented this hack, but it would probably be much cleaner and more reliable to write this code inside the compositor rather than as a shell script.

I recognize YaLTeR's concerns here:

> The consume-window-into-column does not accept an id atm, and I'm not sure it should, because it's a column-level action, not a window-level action.

However, I think it would be generally valuable as a feature. Again, no pressure to implement this.

**Reply by og900aero** [reply]
So, the Swallowing function won't be implemented as factory? It's really useful, though. Anyway, I used it with other WM's.

**Comment by jrvieira** [comment]
Besides media players and other graphical windows spawned from terminals, i use foot terminal pipes for editing/copying scrollback content in a text editor like vim, where it becomes awkward when the child window doesn't exactly replace the space previously occupied by the parent. Spawning a child window tabbed in the same column would solve this elegantly.

There's a lot of value in being able to open windows in the current column and this is just another use case i can think of. For this reason i reenforce this suggestion.

**Comment by Asthestarsfalll** [comment]
Try [this](https://github.com/Asthestarsfalll/piri/blob/main/docs/en/plugins/swallow.md)
Pid matching:
Traces the child process's process tree. For example, mpv -> fish -> ghostty -> system, find the ghostty.

https://github.com/user-attachments/assets/82758092-1a4e-4cfc-9fde-3372b6ae5b5b

Rule based:
Parent Window Discovery Mechanism:
- If the currently focused window is not the child window, use the currently focused window as the candidate parent window
- If the currently focused window is the child window itself, search for a matching parent window from the focus window queue (maintains the last 5 focused windows)
- The focus window queue is automatically updated when windows gain focus

Config:
```toml
[piri.swallow]
use_pid_matching = false

[[swallow]]
child_app_id = '.*google-chrome.*'
parent_app_id = '.*ghostty.*'
```

https://github.com/user-attachments/assets/48c8f9ab-2ec0-403a-8c04-b37dc1ca9dad

**Reply by og900aero** [reply]
Your cursor's flame effect is brilliant! How can u do that?

**Reply by arthsmn** [reply]
> Your cursor's flame effect is brilliant! How can u do that?

Ghostty can use shaders, so you just get it from the internet and declare in your config.

**Reply by Asthestarsfalll** [reply]
> Your cursor's flame effect is brilliant! How can u do that?

https://github.com/Asthestarsfalll/dotfiles/tree/main/.config/ghostty/shaders

**Reply by ArtemChandragupta** [reply]
This works really good
